packages/extension-base/src/background/handlers/Extension.ts

```diff
@@ -287,33 +303,41 @@ export default class Extension {
     };
   }
 
-  private signingApprovePassword ({ id, isSavedPass, password }: RequestSigningApprovePassword): boolean {
+  private signingApprovePassword ({ id, password, savePass }: RequestSigningApprovePassword): boolean {
     const queued = this.#state.getSignRequest(id);
 
     assert(queued, 'Unable to find request');
 
     const { reject, request, resolve } = queued;
-    const pair = keyring.getPair(request.payload.address);
+    const pair = keyring.getPair(queued.account.address);
+
+    // unlike queued.account.address the following
+    // address is encoded with the default prefix
+    // which what is used for password caching mapping
+    const { address } = pair;
 
     if (!pair) {
       reject(new Error('Unable to find pair'));
 
       return false;
     }
 
-    const now = Date.now();
+    this.refreshAccountPasswordCache(pair);
+
+    // if the keyring pair is locked, the password is needed
+    if (pair.isLocked && !password) {
+      reject(new Error('Password needed to unlock the account'));
```

Comment:
BTW, unlike what I said, I actually can't give a custom error here because the rejection does not only occur when the cache expires, it could also simply be an empty password (although this shouldn't be possible any more since I disabled the button).